### PR TITLE
feat(stepfunctions-tasks): support for APIGW API: Invoke

### PR DESCRIPTION
**Pull request by Sumeet-Badyal**
feat(stepfunctions-tasks): support for APIGW API: Invoke

Implementation

Update package @aws-cdk/aws-stepfunctions-tasks to include support for ApiGateway Invoke 
API as per documentation here:
https://docs.aws.amazon.com/step-functions/latest/dg/connect-api-gateway.html
https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html

Includes support for the following Amazon Athena API calls:
Invoke

Closes #11565

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by shivlaks** [comment]
@pahud feel free to give the open PR a review as well!

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (pahud) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:16]: how about 

```ts
readonly api: apigateway.IRestAPI;
```

And maybe generate the `apiEndpoint`  like this?

```ts
const apiStack = cdk.Stack.of(props.api);
this.apiEndpoint = ${props.api.restApiId}.execute-api.${apiStack.region}.${apiStack.urlSuffix}
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (pahud) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:63]: Maybe we should have another enum for this like

```ts
export enum AuthType {
  NO_AUTH = 'NO_AUTH',
  IAM_ROLE = 'IAM_ROLE',
  RESOURCE_POLICY = 'RESOURCE_POLICY',
}
```

and make `AuthType.NO_AUTH` the default?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (pahud) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:97]: If the `api` property is `apigateway.IRestAPI`, we can construct the resource ARN this way:

```ts
this.taskPolicies = [
      new iam.PolicyStatement({
        resources: [this.executeApiArn],
        actions: ['ExecuteAPI:Invoke'],
      }),
    ];
  }
  get executeApiArn() {
    return this.props.api.arnForExecuteApi(this.props.method, this.props.path, this.props.stage)
  }
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (pahud) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:139]: If we've built 

```ts
this.apiEndpoint = ${props.api.restApiId}.execute-api.${apiStack.region}.${apiStack.urlSuffix}
```

we can simply 

```ts
ApiEndpoinit: this.apiEndpoint
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:88]: Where is this value coming from? Is this implicit for the complete state-machine?

```ts
sfn.TaskInput.fromDataAt('$.AuthType').value;
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:90]: We can move the resource generation logic to a private helper which can potentially be used in the `RESOURCE_POLICY` case as well with some arguments.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:134]: The current implementation of this method is not so clean and readable. There is a lot of duplication which we can potentially avoid generating the parameters before we do a return.

Could you please explain the significance of all the different cases in this method and why it needs to be implemented the way it is?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:141]: The same question as before (and at all the places). We have already checked for `this.props.header` in the `if` condition. Why do we need to check it again here? Also, where is this value coming from?

```ts
sfn.TaskInput.fromDataAt('$.Headers')
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:242]: Do you think we should depend on api-gateway module and get this enum from there?

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/invoke.test.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/invoke.test.ts:5]: Please add test cases for the situations where we throw in the constructor.

**Review by ayush987goyal — CHANGES_REQUESTED** [review]
Thanks for this change! Some comments.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:134]: I am thinking why can't we do it this way:

```ts
protected _renderTask(): any {
	return {
        Resource: integrationResourceArn('apigateway', 'invoke', this.integrationPattern),
        Parameters: sfn.FieldUtils.renderObject({
          ApiEndpoint: this.props.apiEndpoint,
          Method: this.props.method,
          Headers: this.props.headers ?? this.props.headers.value,
          Stage: this.props.stage,
          Path: this.props.path,
          QueryParameters: this.props.queryParameters ?? this.props.queryParameters.value,
          RequestBody: this.props.requestBody ?? this.props.requestBody.value,
          AuthType: this.props.authType ?? this.props.authType.value : 'NO_AUTH',
        }),
      };
}
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:16]: this suggestion is essentially what I had in mind too

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:242]: I believe this exists in `apigatewayv2` but not the `apigateway` module... it might also be duplicated in `s3` (or some module, I know it's duplicated elsewhere).

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:245]:
```suggestion
  /** Retrieve data from a server at the specified resource */
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:250]: nit: inline all of these single line doc strings (we only recently started doing this so ignore the prior art within this module)

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:27]: does this need to be a `TaskInput`? - is that because they can be passed through the state input as too?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:33]: I believe this is called `stageName` in the `@aws-cdk/aws-apigateway` module - we should align rather than maintain the naming from the Api

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:24]: I think we can provide a more informative doc string here. maybe throw in an example too?

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: just confirming: this is not the same as `apigatewayv2` correct?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:42]: an example might communicate this better

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:59]: this looks it would be more suitable to be an enum. Any reason why we would need it to be of type `TaskInput` aside from the type in the Sfn documentation stating it supports `Json`?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:88]: curious about this myself... state input isn't guaranteed to have anything in it as the previous state might just provide a `{}`

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:90]: would also appreciate a comment (with an example perhaps) that describes the logic as this will be pretty cryptic to new contributors

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:124]: there's some duplication in this control flow block from the if statement that we can simplify

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:236]: there's a lot of duplicated code in this control flow chain from the `if` on down. I think we can simplify this to be a lot more succinct. let's refactor this please!

**Review by shivlaks — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: Not the same.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:16]: Implementing now. this cleans it up nicely, thanks for the suggestion.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:27]: It's just a json. I can change the type.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:97]: I am going to rename the method to arnForExecuteApi as it was renamed to that in aws-apigateway

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:242]: The one in s3 does not have option method

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:20]: I think this should not be needed anymore.

**Review by ayush987goyal — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:108]:
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:128]: nit: I think the doc string should be a little more informative than the method name. Perhaps describe what that ARN is / what it's used for, show an example of what it looks like, etc.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:207]: enums don't have defaults, it's only where they're used that would
```suggestion
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts:11]: > should return the same response as the state machine

nit: better to be explicit and clear about what that response is

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts:28]: does this need to be a function? - why not just define it ahead of it's usage or inline it?

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts:11]: should the rest-api-id also be an output? same question for all the inputs here. steps should include where they retrieved from. you might need to add outputs to the stack if they're not easily available.

**Review by shivlaks — CHANGES_REQUESTED** [review]
would also love if @nija-at could take a look since he's the primary maintainer of the `api-gateway` module

File: packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/test/apigateway/integ.invoke.ts]: Could you clarify what you mean by this? I am not sure how to get the rest-api-id.

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: But this works to invoke functions defined via apigatewayv2, correct?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: Anyone digging into this? Happy to try and help; looking forward to putting this functionality to use ;)

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: haven't played with it myself yet, can you clarify @Sumeet-Badyal

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: @eikeon - happy to have your help here!! :)

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (nija-at) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: From this page - https://docs.aws.amazon.com/step-functions/latest/dg/connect-api-gateway.html

> You use Amazon API Gateway to create, publish, maintain, and monitor HTTP and REST APIs. To integrate with API Gateway, you define a Task state in Step Functions that directly calls an API Gateway HTTP or API Gateway REST endpoint, without writing code or relying on other infrastructure.

This task supports invoking Rest APIs (v1) and HTTP APIs (v2) but not websocket APIs (v2).

Given our modeling of higher level constructs, we'll need to implement this as two tasks (not necessarily in the same PR). One that takes `IRestApi` from the `@aws-cdk/aws-apigateway` module and another that takes `IHttpApi` from the `@aws-cdk/aws-apigatewayv2` module.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts
Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: Is this pretty much the sole remaining thing holding up this pull request? Can save some time to attempt to dig into this headers issue later today

Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: headers here probably should be similar to input here: packages/@aws-cdk/aws-stepfunctions-tasks/test/start-execution.test.ts and packages/@aws-cdk/aws-stepfunctions-tasks/lib/start-execution.ts -- and they both have the same type. Gotta run for a few hours, but will trace how headers is being used in this pull request compared to other working examples.

Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: And https://docs.aws.amazon.com/step-functions/latest/dg/connect-api-gateway.html

Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: readonly headers?: { [key: string]: string[] }

Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/apigateway/invoke.ts:30]: I also see a string value in one of the examples on that doc page: readonly headers?: { [key: string]: string[] | string }

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md
Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: Ah; looks like someone still needs to add IHttpApi too

Comment (eikeon) [reply packages/@aws-cdk/aws-stepfunctions-tasks/README.md:218]: @Sumeet-Badyal are you still working on this? Any plans to add the second task for the IHttpApi case? Else when my aws-cdk dev container finishes spinning up I could have a go -- imagine there'd be common code between the two we'd want to abstract into a shared place in the end

**Comment by mmcircle** [comment]
Hi Guys, what's up with the feature ?

**Comment by shivlaks** [comment]
superseded by #13033 

@mmcircle - efforts to land this have been resumed in a different PR (#13033). please follow that one for the current status of API Gateway service integration support.

**Pull request closed without merging**
